Content.Server/Abilities/Psionics/PsionicAbilitiesSystem.cs

```diff
@@ -62,66 +77,185 @@ public void AddPsionics(EntityUid uid, bool warn = true)
-                _sawmill.Error("Can't index the random psionic power pool!");
-                return;
+                EntityUid? actionId = null;
+                if (_actions.AddAction(uid, ref actionId, id))
+                {
+                    _actions.StartUseDelay(actionId);
+                    psionic.Actions.Add((id, actionId));
+                }
             }
 
-            // uh oh, stinky!
-            var newComponent = (Component) _componentFactory.GetComponent(pool.Pick());
-            newComponent.Owner = uid;
+            if (proto.AmplificationModifier != 0)
+                psionic.AmplificationSources.Add(proto.Name, proto.AmplificationModifier);
 
-            EntityManager.AddComponent(uid, newComponent);
+            if (proto.DampeningModifier != 0)
+                psionic.DampeningSources.Add(proto.Name, proto.DampeningModifier);
 
-            _glimmerSystem.Glimmer += _random.Next(1, 5);
+            if (playPopup)
+                _popups.PopupEntity("generic-power-initialization-feedback", uid, uid, PopupType.MediumCaution);
+            // TODO: Replace this with chat message: _popups.PopupEntity(proto.InitializationFeedback, uid, uid, PopupType.MediumCaution);
+
+            if (proto.Components is not null)
+                foreach (var comp in proto.Components)
+                    if (!EntityManager.TryGetComponent(uid, _componentFactory.GetComponent(comp).GetType(), out var powerComp)
+                        && powerComp is not null)
+                        AddComp(uid, powerComp);
+
+            RefreshPsionicModifiers(uid, psionic);
+        }
+
+        /// <summary>
+        ///     Updates a Psion's casting stats, call this anytime a system adds a new source of Amp or Damp.
+        /// </summary>
+        /// <param name="uid"></param>
+        /// <param name="comp"></param>
+        public void RefreshPsionicModifiers(EntityUid uid, PsionicComponent comp)
+        {
+            var ampModifier = 0f;
+            var dampModifier = 0f;
+            foreach (var (_, source) in comp.AmplificationSources)
+                ampModifier += source;
+            foreach (var (_, source) in comp.DampeningSources)
+                dampModifier += source;
+
+            var ev = new OnSetPsionicStatsEvent(ampModifier, dampModifier);
+            RaiseLocalEvent(uid, ref ev);
+            ampModifier = ev.AmplificationChangedAmount;
+            dampModifier = ev.DampeningChangedAmount;
+
+            comp.CurrentAmplification = ampModifier;
+            comp.CurrentDampening = dampModifier;
         }
 
-        public void RemovePsionics(EntityUid uid)
+        /// <summary>
+        ///     Updates a Psion's casting stats, call this anytime a system adds a new source of Amp or Damp.
+        ///     Variant function for systems that didn't already have the PsionicComponent.
+        /// </summary>
+        /// <param name="uid"></param>
+        public void RefreshPsionicModifiers(EntityUid uid)
```

Comment:
This function is identical to the one above it, except for the parameters. It should be possible to make the latter call the first.

```
public void RefreshPsionicModifiers(EntityUid uid)
{
    if (!TryComp<PsionicComponent>(uid, out var comp))
                return;
                
    RefreshPsionicModifiers(uid, comp);
}
```
This will greatly reduce the amount of redundant code, making it a lot easier to maintain.